Add Folder Lint Option to Right Click Folder Action Menu

pjkaufman opened the pull request:
@platers 

Fixes #177 

I essentially duplicated the confirmation model and changed a few things as well as making sure that the event action would be added to the drop down menu when your right click a folder.

I have a couple of questions regarding how things were done here and whether I should go ahead and try to change a few things:
- Is there a more performant way to determine which files are in the folder selected or do I just have to iterate over all files and just see if they start with the same path as the folder selected?
- Can the modal code be made generic to allow for the addition of a couple of extra params to cut down on the duplication of code? I was finding it hard to pass in a function as a param.
- Do we want to try to account for the wording that would be needed for folders with only 1 markdown file in it?
- Do we want the confirmation modal for the request to lint the files in the folder?
- Is it logical to assume that linting the folder will lint all subfolders as well?
- Are there any documentation updates or UT updates that are needed?

Thanks for the guidance on this. I am hoping this will benefit those that use the plugin and hopefully be straight forward in the code.

platers commented:
Thanks for the PR! I'm checking it out now. I'm having trouble finding the button is to lint folders, can you send a screenshot? Also it would be nice to have a command to lint files in the current folder.

pjkaufman commented:
> Thanks for the PR! I'm checking it out now. I'm having trouble finding the button is to lint folders, can you send a screenshot? Also it would be nice to have a command to lint files in the current folder.

You need to right click on the folder to get it to to show up in the list. As far as adding the command goes, that should be relatively easy since it will just require a refactor to get the linting code reusable from both places. 

Here is a screenshot:
![Screenshot from 2022-05-12 17-34-59](https://user-images.githubusercontent.com/29898683/168172151-424d56d8-ab64-4185-9f45-558da49b8fa6.png)

pjkaufman commented:
The command is now present.

pjkaufman commented:
Please let me know if the command seems to act weird.

platers commented on src/main.ts:
```diff
@@ -180,6 +204,18 @@ export default class LinterPlugin extends Plugin {
+      console.log('Linting folder ' + folder.name);
+      let lintedFiles = 0;
+      await Promise.all(this.app.vault.getMarkdownFiles().map(async (file) => {
+        if (file.path.startsWith(folder.path) && !this.shouldIgnoreFile(file)) {
```
when I lint the folder named `Test` it also lints files starting with `Test` like Testing.md which it shouldnt

pjkaufman replied:
This should be taken care of with the latest commit.

platers commented:
> @platers
> 
> Fixes #177
> 
> I essentially duplicated the confirmation model and changed a few things as well as making sure that the event action would be added to the drop down menu when your right click a folder.
> 
> I have a couple of questions regarding how things were done here and whether I should go ahead and try to change a few things:
> 
> * Is there a more performant way to determine which files are in the folder selected or do I just have to iterate over all files and just see if they start with the same path as the folder selected?

I dont know a better way

> * Can the modal code be made generic to allow for the addition of a couple of extra params to cut down on the duplication of code? I was finding it hard to pass in a function as a param.

Yeah it would be nice to reuse code there. Maybe a more generic Modal class? I'm ok with leaving it as is though.

> * Do we want to try to account for the wording that would be needed for folders with only 1 markdown file in it?

Why are folders with 1 file special?

> * Do we want the confirmation modal for the request to lint the files in the folder?

I think its a good idea since users wont be able to see the files being linted when they run it.

> * Is it logical to assume that linting the folder will lint all subfolders as well?

Makes sense.

> * Are there any documentation updates or UT updates that are needed?

Could you document this feature to the README? Just a sentence or two in the usage section. 

> 
> Thanks for the guidance on this. I am hoping this will benefit those that use the plugin and hopefully be straight forward in the code.

Thanks for the contribution, it looks basically good to me! Just noticed one bug that should be a quick fix

pjkaufman commented:
> > @platers
> > Fixes #177
> > I essentially duplicated the confirmation model and changed a few things as well as making sure that the event action would be added to the drop down menu when your right click a folder.
> > I have a couple of questions regarding how things were done here and whether I should go ahead and try to change a few things:
> > 
> > * Is there a more performant way to determine which files are in the folder selected or do I just have to iterate over all files and just see if they start with the same path as the folder selected?
> 
> I dont know a better way
> 
> > * Can the modal code be made generic to allow for the addition of a couple of extra params to cut down on the duplication of code? I was finding it hard to pass in a function as a param.
> 
> Yeah it would be nice to reuse code there. Maybe a more generic Modal class? I'm ok with leaving it as is though.
> 
> > * Do we want to try to account for the wording that would be needed for folders with only 1 markdown file in it?
> 
> Why are folders with 1 file special?
> 
> > * Do we want the confirmation modal for the request to lint the files in the folder?
> 
> I think its a good idea since users wont be able to see the files being linted when they run it.
> 
> > * Is it logical to assume that linting the folder will lint all subfolders as well?
> 
> Makes sense.
> 
> > * Are there any documentation updates or UT updates that are needed?
> 
> Could you document this feature to the README? Just a sentence or two in the usage section. Make sure to add it in the readme_template and not the README.md since its generated.
> 
> > Thanks for the guidance on this. I am hoping this will benefit those that use the plugin and hopefully be straight forward in the code.
> 
> Thanks for the contribution, it looks basically good to me! Just noticed one bug that should be a quick fix

The special case I was referring to in with 1 file is that when it gets done linting it will say "All 1 files in FOLDER_NAME". I am not sure if there is a different set of wording we would like.

However, I was able to make the modal generic and be used by both.

I updated the README template to include the info and to update the contribution steps to include the mentioning of forking the repo since one cannot update this repo directly by default.

platers reviewed (approved):
LGTM
